kubernetes/deployment/controller.go

```diff
@@ -0,0 +1,282 @@
+	// a second
+	wait.Until(c.runWorker, time.Second, stopCh)
+}
+
+func (c *Controller) runWorker() {
+	// processNextWorkItem automatically waits for work
+	for c.processNextItem() {
+		// loop
+	}
+}
+
+func (c *Controller) processNextItem() bool {
+	// pull next key from queue. Look up key in cache.
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	// Indicate to queue key has been processed
+	defer c.queue.Done(key)
+
+	// do processing on key
+	err := c.processItem(key.(string))
+
+	if err == nil {
+		// No error => queue stop tracking history
+		c.queue.Forget(key)
+	} else if c.queue.NumRequeues(key) < maxRetries {
+		c.Error("processing queue item, will retry", "key", key, "error", err)
+		// requeue
+		c.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		c.Error("processing queue item failed", "key", key, "error", err)
+		c.queue.Forget(key)
+		utilruntime.HandleError(err)
+	}
+
+	return true
+}
+
+// processItem(key) is where we define how to react to an item coming off the
+// work queue.
+func (c *Controller) processItem(key string) error {
+	c.Debug("processing change to pod", "pod", key)
+
+	obj, exists, err := c.informer.GetIndexer().GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("Error fetching object with key %s from store: %v", key, err)
+	}
+
+	if !exists {
+		c.Debug("object deleted", "key", key, "obj", obj)
+		return nil
+	}
+	err = c.processObj(obj)
+	return err
+}
+
+// Process a generic object.
+func (c *Controller) processObj(obj interface{}) error {
+	switch v := obj.(type) {
+	case *apiv1.Pod:
+		c.processPod(v)
+	default:
+		return error(fmt.Errorf("obj is not a pod"))
+	}
+	return nil
+}
+
+// processPod defines how to react to a pod coming off the workqueue in an
+// observed state. Assumes there is only 1 initcontainer.
+func (c *Controller) processPod(obj *apiv1.Pod) {
+	if len(obj.Status.InitContainerStatuses) != 0 {
+		switch {
+		case obj.Status.InitContainerStatuses[0].State.Waiting != nil:
+			c.Debug("init container waiting!")
+
+		case obj.Status.InitContainerStatuses[0].State.Running != nil:
+			c.Debug("init container running!")
+			c.Debug("calling CopyTar", "files", c.Opts.Files)
+			err := c.Client.CopyTar(c.Opts.Files, obj)
+			if err != nil {
+				c.Error("copying tarball", "error", err)
+			}
+		case obj.Status.ContainerStatuses[0].State.Running != nil:
+			// Write the pod name, name to the resources file. This allows us to
+			// retrieve it to obtain the client.token
+			resources := &cloud.Resources{}
+			file, err := os.OpenFile(c.Opts.ResourcePath, os.O_RDONLY, 0600)
+			if err != nil {
+				c.Error("could not open resource file", "path", c.Opts.ResourcePath, "error", err)
+			}
+			c.Debug("opened resource file", "path", c.Opts.ResourcePath)
+			decoder := gob.NewDecoder(file)
+			err = decoder.Decode(resources)
+			if err != nil {
+				c.Error("decoding resource file", "error", err)
```

Comment:
Is it intended that we just continue on with an empty resources struct on failure to Decode?